<a href="https://colab.research.google.com/github/tsingjinyun/----------/blob/master/NLP_%E5%85%A5%E9%96%80_(1_3)_Word2Vec_Sentiment_Analysis.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

### 準備 Pre-trained 的 Word2Vec weights  
歡迎參考我的這篇[blog](https://medium.com/@sfhsu29/nlp-%E5%B0%88%E6%AC%84-1-2-%E5%A6%82%E4%BD%95%E8%A8%93%E7%B7%B4%E8%87%AA%E5%B7%B1%E7%9A%84-word2vec-5a0754c5cb09)來使用維基百科的資料訓練自己的word vector  
如果只想下載 word2vec 參數到你的local的話：[google drive連結](https://docs.google.com/u/0/uc?export=download&confirm=8s0o&id=17ThJchrSwnOy1HEKX1HT79vpl1y4UH2L)

In [ ]:
!wget --load-cookies /tmp/cookies.txt "https://docs.google.com/uc?export=download&confirm=$(wget --quiet --save-cookies /tmp/cookies.txt --keep-session-cookies --no-check-certificate 'https://docs.google.com/uc?export=download&id=17ThJchrSwnOy1HEKX1HT79vpl1y4UH2L' -O- | sed -rn 's/.*confirm=([0-9A-Za-z_]+).*/\1\n/p')&id=17ThJchrSwnOy1HEKX1HT79vpl1y4UH2L" -O wiki-word2vec.zip && rm -rf /tmp/cookies.txt
!unzip wiki-word2vec.zip

--2020-12-27 00:16:04--  https://docs.google.com/uc?export=download&confirm=AOlH&id=17ThJchrSwnOy1HEKX1HT79vpl1y4UH2L
Resolving docs.google.com (docs.google.com)... 74.125.202.138, 74.125.202.102, 74.125.202.100, ...
Connecting to docs.google.com (docs.google.com)|74.125.202.138|:443... connected.
HTTP request sent, awaiting response... 302 Moved Temporarily
Location: https://doc-0s-64-docs.googleusercontent.com/docs/securesc/sulptmrivnfhk17i3c0jcdn7davgj883/4o4anhefs66ofud344tis43flkd54jjc/1609028100000/18293686574256102848/03278925523020089386Z/17ThJchrSwnOy1HEKX1HT79vpl1y4UH2L?e=download [following]
--2020-12-27 00:16:04--  https://doc-0s-64-docs.googleusercontent.com/docs/securesc/sulptmrivnfhk17i3c0jcdn7davgj883/4o4anhefs66ofud344tis43flkd54jjc/1609028100000/18293686574256102848/03278925523020089386Z/17ThJchrSwnOy1HEKX1HT79vpl1y4UH2L?e=download
Resolving doc-0s-64-docs.googleusercontent.com (doc-0s-64-docs.googleusercontent.com)... 172.217.212.132, 2607:f8b0:4001:c03::84
Connectin

#### 使用 Gensim 來 Load 剛剛下載好的參數

In [ ]:
from gensim.models import Word2Vec
model = Word2Vec.load("wiki-lemma-100D-phrase")

#### Google 的 word representation, word embedding, word vector (這三這幾乎是同義詞)

In [ ]:
model.wv['Google']

array([ 3.3885376 , -4.2511983 , -1.445084  ,  1.9302932 ,  0.84648174,
       -0.63184077,  0.05726242,  0.7063744 , -0.05059653,  2.1846352 ,
       -0.12328646,  0.6471171 ,  0.25026178, -0.14930539, -0.07820829,
        0.16145864,  0.6498362 ,  1.0886047 ,  1.5343726 ,  4.950689  ,
        0.50769436, -0.3946989 ,  2.1025164 , -3.1834273 ,  2.380777  ,
        0.24872196, -2.5802555 ,  1.4149348 ,  0.63408375, -1.6591691 ,
       -2.4389815 ,  2.1249    ,  0.39783332, -0.35228804, -2.9626958 ,
       -4.475956  , -4.0608654 , -0.39374784, -0.37150937, -1.4930124 ,
       -2.4813695 ,  2.203496  ,  1.9972353 ,  5.479271  , -0.00601585,
        2.1382024 , -1.6378103 ,  1.9081411 , -0.6341882 , -2.2414782 ,
       -0.70867395,  1.1144645 , -1.7722839 ,  0.37062666,  0.80082476,
        1.9306395 ,  0.48852623, -1.1423063 ,  3.0780652 ,  0.7570495 ,
       -2.8385687 ,  0.5158689 , -1.316193  ,  0.51292163,  1.862376  ,
        1.7524154 ,  1.8987881 ,  0.06306809,  1.378573  , -3.18

和Google 在向量空間裡，最靠近（cosine similarity）的單詞們！  
在訓練時我是使用case-sensitive且包含片語的

In [ ]:
model.wv.most_similar('Google', topn = 5)

[('Microsoft', 0.8559668064117432),
 ('Google_Search', 0.8424477577209473),
 ('Google_Analytics', 0.8412378430366516),
 ('Yahoo', 0.8326560854911804),
 ('AltaVista', 0.8248958587646484)]

### 2. 下載 sentiment analysis 所要用到的 IMDB 資料集，並解壓縮到colab drive裡面

In [ ]:
! wget http://ai.stanford.edu/~amaas/data/sentiment/aclImdb_v1.tar.gz
! tar zxf aclImdb_v1.tar.gz

--2020-12-27 00:16:14--  http://ai.stanford.edu/~amaas/data/sentiment/aclImdb_v1.tar.gz
Resolving ai.stanford.edu (ai.stanford.edu)... 171.64.68.10
Connecting to ai.stanford.edu (ai.stanford.edu)|171.64.68.10|:80... connected.
HTTP request sent, awaiting response... 200 OK
Length: 84125825 (80M) [application/x-gzip]
Saving to: ‘aclImdb_v1.tar.gz’

aclImdb_v1.tar.gz   100%[===================>]  80.23M  25.0MB/s    in 4.3s    

2020-12-27 00:16:19 (18.7 MB/s) - ‘aclImdb_v1.tar.gz’ saved [84125825/84125825]



#### 從解壓縮的資料夾讀取 IMDB 資料

In [ ]:
import os

def read_files_to_list(path):
  docs = []
  # 借助os.listdir找出特定folder下所有的files
  files = os.listdir(path)
  for file in files:  
    # 再把path 和 file names join起來，就可以得到我們要的檔案位置
    with open(os.path.join(path, file)) as f:
      docs.append(f.read())
 
  return docs

In [ ]:
# Set the path of positive examples
path_pos = "aclImdb/train/pos"
# Set the path of negative examples
path_neg = "aclImdb/train/neg"

pos_file_list = read_files_to_list(path_pos)
neg_file_list = read_files_to_list(path_neg)

In [ ]:
 # 總共有25000個 training data
 len(pos_file_list+neg_file_list)

25000

In [ ]:
import pandas as pd
# 把 positive 的文章list 和 negative 的文章 list 串接再一起，在和他們對應的label zip再一起，變成 
# ((positive article, 1)
#  (positive article, 1)
#  ...
#  (negative article, 0))
imdb_df = pd.DataFrame(data = zip(pos_file_list +neg_file_list, [1] * len(pos_file_list) + [0] * len(neg_file_list)))
imdb_df.columns = ['text', 'label']

In [ ]:
imdb_df.head()

text  label
0  Yesterday, I went to the monthly Antique Flea ...      1
1  I watched "Elephant Walk" for the first time i...      1
2  You probably all already know this by now, but...      1
3  I love this movie/short thing. Jason Steele is...      1
4  Great movie, enough laughs and action for any ...      1

In [ ]:
imdb_df.tail()

text  label
24995  If you like to comment on films where the scri...      0
24996  The social commentary was way overblown and th...      0
24997  In the first Howling, we are introduced to a w...      0
24998  I loved Adrianne Curry before this show. I tho...      0
24999  <br /><br />I am a big-time horror/sci-fi fan ...      0

#### 在我第一篇[blog](https://sfhsu29.medium.com/nlp-%E5%85%A5%E9%96%80-1-text-classification-sentiment-analysis-%E6%A5%B5%E7%B0%A1%E6%98%93%E6%83%85%E6%84%9F%E5%88%86%E9%A1%9E%E5%99%A8-bag-of-words-naive-bayes-e40d61de9a7f) 之中提到，如果使用：
- one-hot encoding 且沒有 remove stop words: 0.849
- one-hot encoding 且 remove stop words: 0.859
- tf-idf encoding 且沒有 remove stop words: 0.864
- tf-idf encoding 且 remove stop words: 0.867

讓我們來測試看看 word embedding 有沒有更厲害

### Baseline: 將每個單詞的詞向量取平均，在使用 Support Vector Classifier 來分類 (0.802 accuracy)

(想要使用 Deep Learning 的話，請直接跳到下一個單元)

#### 1. 清理資料：
當我們訓練word2vec的時候，我們選擇了Lemmatize了wikipedia的資料，因此我們也要用相同的步驟來清理IMDB的資料

In [ ]:
import spacy
import spacy.cli
spacy.cli.download("en_core_web_md")

In [ ]:
nlp = spacy.load('en_core_web_md', disable=["ner", "parser"])

#### 2. 處理文字：


1.   使用 spacy 的 lemma_ 來取得單詞 lemmatize 的結果
2.   使用 w2vmodel.wv[word.lemma_] 來取得該單詞的 word embeddings 最後再將整個句子的 word embeddings 取平均。



In [ ]:
# 示範一下 lemma 的結果
for token in  nlp("Apple is looking at buying U.K. startup for $1 billion"):
  print('Word: {}  --> {}'.format(token.text, token.lemma_))

In [ ]:
import numpy as np
# Get all words from wiki word2vec model
model_words = set(model.wv.index2word)

def text2vec_raw(w2vmodel, text):
  doc = nlp(text)
  # convert a movie review into vectors
  text_vecs = [w2vmodel.wv[word.lemma_] for word in doc if word.lemma_ in model_words]
  # calculate the mean of the vectors and return
  if len(text_vecs) > 1:
    res =  np.mean(text_vecs, axis = 0)
    return res
  elif len(text_vecs) == 1:
    return res[0]
  else:
    return np.nan


In [ ]:
imdb_df['vectors'] = imdb_df['text'].apply(lambda x: text2vec_raw(model, x))

In [ ]:
imdb_df

In [ ]:
# 找出是否有 label 不為 pos, neg (沒有)
imdb_df[~imdb_df['label'].isin(['pos', 'neg'])]

In [ ]:
vector_df = pd.DataFrame(imdb_df.vectors.tolist())
vector_df['label'] = imdb_df['label'].values

In [ ]:
# 使用 Sample 來 shuffle data
vector_df = vector_df.sample(frac =1.0)

#### 3. 建立一個 model 來預測句子的sentiment (scikit learn, Machine Learning)


1.   RandomForestClassifier = 0.757 (可能略有不同)
2.   GradientBoostingClassifier = 0.772 (可能略有不同)
3.   SVM classifier = 0.801 (可能略有不同)

-> 可以發現模型的正確率並沒有比 Bag-of-words的方法來的好  
-> 錯誤分析: 在之前的步驟之中，我們假設 sentence embedding = average of word embeddings，這個取平均的動作，會讓 embeddings 之間的「特性」被互相抵消。



#### 1. Random Forest Classifier

In [ ]:
from sklearn.model_selection import cross_val_score
from sklearn.ensemble import RandomForestClassifier

In [ ]:
clf = RandomForestClassifier()
Y = vector_df['label'].values
X = vector_df.drop(['label'], axis = 1)
 
scores = cross_val_score(clf, X, Y, cv = 3)

In [ ]:
print(scores.mean())

#### 2. Gradient Boosting Classifier

In [ ]:
from sklearn.ensemble import GradientBoostingClassifier

In [ ]:
clf = GradientBoostingClassifier()
Y = vector_df['label'].values
X = vector_df.drop(['label'], axis = 1)
 
scores = cross_val_score(clf, X, Y, cv = 3)

In [ ]:
print(scores.mean())

#### 3. Support Vector Classifier

In [ ]:
from sklearn.svm import SVC
clf = SVC()
Y = vector_df['label'].values
X = vector_df.drop(['label'], axis = 1)
 
scores = cross_val_score(clf, X, Y, cv = 3)

In [ ]:
print(scores.mean())

### 建立一個 model 來預測句子的 sentiment (keras, deep learning)

In [ ]:
# 由於我們剛剛下載的 word2vec model 是專門給 gensim 用的，如果我們要給 keras 使用，我們要先將model儲存成不同的格式
model.wv.save_word2vec_format('keras_word2vec.txt', binary=False)

#### 將 word embeddings 儲存到字典(dictionary)裡面

In [ ]:
import numpy as np
# load the whole embedding into memory
embeddings_index = dict()
with open('keras_word2vec.txt') as f:
  for line in f:
    values = line.split()
    # 只接受長度為 101 的向量 (word + 100d embedding)
    if len(values) != 101:
      print(values)
      continue
    word = values[0]
    coefs = np.asarray(values[1:], dtype='float32')
    embeddings_index[word] = coefs

['346848', '100']


#### 找出適合的句子長度：
(最後使用 max_len = 512)

In [ ]:
import numpy as np
### 找出影評共有幾個 words
sentence_lengths = imdb_df['text'].apply(lambda x: len(x.split()))
quantiles = np.percentile(sentence_lengths, [25, 50, 75])

print('The minimal word count in a movie review: ', min(sentence_lengths))
print('The first quantile (25%) of word count in a movie review:', quantiles[0])
print('The second quantile (50%) of word count in a movie review:', quantiles[1])
print('The third quantile (75%) of word count in a movie review:', quantiles[2])
print('The maximum word count in a movie review: ', max(sentence_lengths))

The minimal word count in a movie review:  10
The first quantile (25%) of word count in a movie review: 127.0
The second quantile (50%) of word count in a movie review: 174.0
The third quantile (75%) of word count in a movie review: 284.0
The maximum word count in a movie review:  2470


#### 處理文字 (lemmatize)，並儲存到review_lines

In [ ]:
from tqdm import tqdm
import spacy.cli
spacy.cli.download("en_core_web_md")
nlp = spacy.load('en_core_web_md', disable=["ner", "parser"])

review_lines = list()
reviews = imdb_df['text'].values.tolist()

for review in tqdm(reviews):   
    doc = nlp(review)
    words = [word.lemma_ for word in doc]
    review_lines.append(words)

✔ Download and installation successful
You can now load the model via spacy.load('en_core_web_md')


100%|██████████| 25000/25000 [16:17<00:00, 25.57it/s]


In [ ]:
import numpy as np
from tensorflow.python.keras.preprocessing.text import Tokenizer
from tensorflow.python.keras.preprocessing.sequence import pad_sequences

# 初始化 tokenizer
# 將文字轉換成 2D integer tensor
tokenizer_obj = Tokenizer()
tokenizer_obj.fit_on_texts(review_lines)
sequences = tokenizer_obj.texts_to_sequences(review_lines)

# 因為model的input必須一樣長，所以要使用padding，句子短於512的都會補0，長於512的會被truncate
word_index = tokenizer_obj.word_index
print('Found %s unique tokens.' % len(word_index))
review_pad = pad_sequences(sequences, maxlen=512)

sentiment =  imdb_df['label'].values
print('Shape of review tensor:', review_pad.shape)
print('Shape of sentiment tensor:', sentiment.shape)


Found 90801 unique tokens.
Shape of review tensor: (25000, 512)
Shape of sentiment tensor: (25000,)


In [ ]:
# 建立一個 look up table，第一個 row 就是第一個單詞的 word embedding
vocab_size = len(word_index) + 1
embedding_matrix = np.zeros((vocab_size, 100))
for word, i in tokenizer_obj.word_index.items():
	embedding_vector = embeddings_index.get(word)
	if embedding_vector is not None:
		embedding_matrix[i] = embedding_vector

In [ ]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
from tensorflow.keras.layers import GRU
from tensorflow.keras.layers import Embedding
from keras.initializers import Constant
import tensorflow as tf


# define model
gru_model = Sequential()
e = Embedding(vocab_size, 100, embeddings_initializer=Constant(embedding_matrix), trainable=False)
gru_model.add(e)
gru_model.add(GRU(64, dropout = 0.2, recurrent_dropout = 0.2))
gru_model.add(Dense(1, activation='sigmoid'))
# compile the model
gru_model.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])
# summarize the model
print(gru_model.summary())

Model: "sequential_2"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
embedding_1 (Embedding)      (None, None, 100)         9080200   
_________________________________________________________________
gru_1 (GRU)                  (None, 64)                31872     
_________________________________________________________________
dense_1 (Dense)              (None, 1)                 65        
Total params: 9,112,137
Trainable params: 31,937
Non-trainable params: 9,080,200
_________________________________________________________________
None


In [ ]:
tensorboard_callback = tf.keras.callbacks.TensorBoard(log_dir="logs")
# fit the model
gru_model.fit(review_pad, sentiment, batch_size=128, epochs=10, validation_split=0.1, verbose=1, callbacks=[tensorboard_callback])
# evaluate the model

Epoch 1/10
176/176 [==============================] - 198s 1s/step - loss: 0.6650 - accuracy: 0.5909 - val_loss: 0.5415 - val_accuracy: 0.7548
Epoch 2/10
176/176 [==============================] - 188s 1s/step - loss: 0.4882 - accuracy: 0.7737 - val_loss: 0.3647 - val_accuracy: 0.8408
Epoch 3/10
176/176 [==============================] - 187s 1s/step - loss: 0.3976 - accuracy: 0.8280 - val_loss: 0.3257 - val_accuracy: 0.8832
Epoch 4/10
176/176 [==============================] - 192s 1s/step - loss: 0.3512 - accuracy: 0.8478 - val_loss: 0.3375 - val_accuracy: 0.8744
Epoch 5/10
176/176 [==============================] - 189s 1s/step - loss: 0.3259 - accuracy: 0.8611 - val_loss: 0.2907 - val_accuracy: 0.8900
Epoch 6/10
176/176 [==============================] - 192s 1s/step - loss: 0.3013 - accuracy: 0.8723 - val_loss: 0.3731 - val_accuracy: 0.8416
Epoch 7/10
176/176 [==============================] - 190s 1s/step - loss: 0.2928 - accuracy: 0.8763 - val_loss: 0.2536 - val_accuracy: 0.8968

In [ ]:
%load_ext tensorboard

In [ ]:
%tensorboard --logdir logs

此单元格的输出内容太大，只能在登录的情况下显示。


實驗一：使用隨機 embedding layer 初始值

In [ ]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
from tensorflow.keras.layers import GRU
from tensorflow.keras.layers import Embedding
from keras.initializers import Constant
import tensorflow as tf


# define model
model = Sequential()
e = Embedding(vocab_size, 100, trainable=False)
model.add(e)
model.add(GRU(64, dropout = 0.2, recurrent_dropout = 0.2))
model.add(Dense(1, activation='sigmoid'))
# compile the model
model.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])
# summarize the model
print(model.summary())

Model: "sequential_1"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
embedding_1 (Embedding)      (None, None, 100)         9080200   
_________________________________________________________________
gru_1 (GRU)                  (None, 64)                31872     
_________________________________________________________________
dense_1 (Dense)              (None, 1)                 65        
Total params: 9,112,137
Trainable params: 31,937
Non-trainable params: 9,080,200
_________________________________________________________________
None


In [ ]:
tensorboard_callback = tf.keras.callbacks.TensorBoard(log_dir="logs")
# fit the model
model.fit(review_pad, sentiment, batch_size=128, epochs=10, validation_split=0.1, verbose=1, callbacks=[tensorboard_callback])
# evaluate the model

Epoch 1/10
176/176 [==============================] - 203s 1s/step - loss: 0.6848 - accuracy: 0.5590 - val_loss: 0.7266 - val_accuracy: 0.3812
Epoch 2/10
176/176 [==============================] - 200s 1s/step - loss: 0.6669 - accuracy: 0.5947 - val_loss: 0.7439 - val_accuracy: 0.4416
Epoch 3/10
176/176 [==============================] - 199s 1s/step - loss: 0.6384 - accuracy: 0.6369 - val_loss: 0.8711 - val_accuracy: 0.1556
Epoch 4/10
176/176 [==============================] - 199s 1s/step - loss: 0.6437 - accuracy: 0.6384 - val_loss: 0.7835 - val_accuracy: 0.4760
Epoch 5/10
176/176 [==============================] - 200s 1s/step - loss: 0.6097 - accuracy: 0.6666 - val_loss: 0.7672 - val_accuracy: 0.4128
Epoch 6/10
176/176 [==============================] - 199s 1s/step - loss: 0.6199 - accuracy: 0.6606 - val_loss: 0.7508 - val_accuracy: 0.5308
Epoch 7/10
176/176 [==============================] - 199s 1s/step - loss: 0.5891 - accuracy: 0.6886 - val_loss: 0.6567 - val_accuracy: 0.6256

In [ ]:
%tensorboard --logdir logs

此单元格的输出内容太大，只能在登录的情况下显示。


In [ ]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
from tensorflow.keras.layers import GRU
from tensorflow.keras.layers import Embedding
from keras.initializers import Constant
import tensorflow as tf


# define model
model = Sequential()
e = Embedding(vocab_size, 100, trainable=True)
model.add(e)
model.add(GRU(64, dropout = 0.2, recurrent_dropout = 0.2))
model.add(Dense(1, activation='sigmoid'))
# compile the model
model.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])
# summarize the model
print(model.summary())

Model: "sequential_2"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
embedding_2 (Embedding)      (None, None, 100)         9080200   
_________________________________________________________________
gru_2 (GRU)                  (None, 64)                31872     
_________________________________________________________________
dense_2 (Dense)              (None, 1)                 65        
Total params: 9,112,137
Trainable params: 9,112,137
Non-trainable params: 0
_________________________________________________________________
None


In [ ]:
tensorboard_callback = tf.keras.callbacks.TensorBoard(log_dir="logs")
# fit the model
model.fit(review_pad, sentiment, batch_size=128, epochs=10, validation_split=0.1, verbose=1, callbacks=[tensorboard_callback])
# evaluate the model

Epoch 1/10
176/176 [==============================] - 248s 1s/step - loss: 0.6363 - accuracy: 0.6111 - val_loss: 0.4450 - val_accuracy: 0.8168
Epoch 2/10
176/176 [==============================] - 241s 1s/step - loss: 0.2860 - accuracy: 0.8822 - val_loss: 0.2818 - val_accuracy: 0.8792
Epoch 3/10
176/176 [==============================] - 242s 1s/step - loss: 0.1641 - accuracy: 0.9394 - val_loss: 0.4890 - val_accuracy: 0.8128
Epoch 4/10
176/176 [==============================] - 243s 1s/step - loss: 0.0822 - accuracy: 0.9733 - val_loss: 0.6543 - val_accuracy: 0.7864
Epoch 5/10
176/176 [==============================] - 242s 1s/step - loss: 0.0502 - accuracy: 0.9848 - val_loss: 0.5528 - val_accuracy: 0.8512
Epoch 6/10
176/176 [==============================] - 243s 1s/step - loss: 0.0337 - accuracy: 0.9896 - val_loss: 0.3610 - val_accuracy: 0.8924
Epoch 7/10
176/176 [==============================] - 243s 1s/step - loss: 0.0314 - accuracy: 0.9899 - val_loss: 0.5734 - val_accuracy: 0.8716